### ISSUE TITLE
Test apps should not use in-process transport

### ISSUE THREAD

**Issue by dsyer** [spring-projects-experimental/spring-grpc#77]
We made the in-process transport the default for tests, but tests with `main()` methods want those to run with the actual TCP socket open. For now you have to manually add `spring.grpc.inprocess.enabled=false` to the main method, but it would be better if we found a way to flip the default for a non-test execution. Is there a way to detect that the application context is being created by spring-test?

**Comment by panic08** [comment]
Can we create a special class that will `implements ApplicationContextInitializer<ConfigurableApplicationContext>` and check the property `spring.test.context`. If it runs with the test, then it will not be null and accordingly we will cast `System.setProperty(“spring.grpc.inprocess.enabled”, false)`

If I understand everything correctly, can you assign me to this issue?

**Comment by dsyer** [comment]
I know you like them but I'm not sure `ApplicationContextInitializer` is the right way to do it. If there is already a system property "spring.test.context" (I was not aware) then we can simply use that in the conditions on the in-process autoconfiguration, can't we?

**Comment by panic08** [comment]
Are you suggesting to just create a configuration class that will have a bin that returns ConfigurableEnvironment and performs the logic I described above, then if all is well, sets `System.setProperty(“spring.grpc.inprocess.enabled”, “false”);` ?

If so, then yes, I think I'd like to do that

**Comment by dsyer** [comment]
That won’t work if junit is on the classpath, which it will be for a test app. There must be a signal that a test is running, I just don’t know what it is yet.

**Comment by panic08** [comment]
Couldn't we just create a Spring test profile for application.properties and throw in `spring.grpc.inprocess.enabled=false`?

**Comment by dsyer** [comment]
I guess that might work, but it seems like that should be a concern of spring-test or spring-boot, so I’m still hoping they already provide something that we don’t know about.

**Comment by dsyer** [comment]
Another option is that we could provide an annotation for tests that want to opt-in to in-process (right now it’s the default based on the classpath).

**Comment by panic08** [comment]
Are you suggesting that we create an annotation that we use to label classes and that will serve as a marker for us to switch spring.grpc.inprocess.enabled to false in this context?

**Comment by dsyer** [comment]
No, the opposite. But I’d prefer to find that there is already a way to know that I am running in a test.

**Comment by panic08** [comment]
I was just thinking. Once again, we want to know we're in the test.

Can we use something like `@ConditionalOnClass(name = “org.springframework.test.context.TestContextManager”)` ?
This class will not exist when running normally, in turn when running the context using test, it will exist

**Comment by dsyer** [comment]
That’s the same as your `org.junit.jupiter.api.Test` suggestion, and it won’t work for the same reason.

**Comment by onobc** [comment]
I am in favor of creating an annotation that a test could be marked with and that would enable the in process server.

There are quite a few places in Spring Boot testing where one can mark a test w/ an annotation to include some other behavior that is otherwise disabled (e.g. [@AutoConfigureObservability](https://docs.spring.io/spring-boot/reference/testing/spring-boot-applications.html#testing.spring-boot-applications.observations)).

I can also imagine in the future having a Spring Boot [test slice](https://docs.spring.io/spring-boot/reference/testing/spring-boot-applications.html#testing.spring-boot-applications.autoconfigured-tests) (e.g. `@GprcTest`) which could include this signal as well.

**Comment by dsyer** [comment]
OK, so let's disable in-process by default just by flipping the default, and add `@AutoConfigureInProcessTransport` that just switches it on. That should be straighforward.

**Comment by panic08** [comment]
Can you assign me?

**Comment by dsyer** [comment]
No need. Don’t wait. It’s not like there are other people going to start on it.

**Comment by DanielLiu1123** [comment]
[grpc-starter](https://github.com/DanielLiu1123/grpc-starter) uses EnvironmentPostProcessor to solve this problem: [GrpcTestEnvironmentPostProcessor](https://github.com/DanielLiu1123/grpc-starter/blob/main/grpc-extensions/grpc-test/src/main/java/grpcstarter/extensions/test/GrpcTestEnvironmentPostProcessor.java).

**Comment by dsyer** [comment]
I like the idea of the annotation though (even if the implementation has an `EnvironmentPostProcessor` in it somewhere - or not). It makes it explicit that you want the in-process transport for the test. Thanks for the suggestion though, and please feel free to make contact out of band if you want to help more.

**Comment by DanielLiu1123** [comment]
Thank you for your response.

I’d like to further explain why `@AutoConfigureInProcessTransport` is unnecessary and might even be a transitional design.

Let’s compare it to a web application. When `@SpringBootTest` works with `spring-boot-starter-web`, the default behavior is to start a mock environment (WebEnvironment.MOCK). This is a very reasonable default and naturally corresponds to gRPC’s in-process transport. Therefore, when testing with the `@SpringBootTest` annotation, there’s no need to explicitly specify using in-process transport because **it should be the default**. If `@AutoConfigureInProcessTransport` didn’t exist, what port would the tests use? User-defined? Random? Or not start at all? None of these options are ideal.

There might also be tests that need to start a real gRPC server, possibly using a user-defined port or a random port (similar to `webEnvironment` in `@SpringBootTest`). How should such cases be handled?

Providing configuration options might be a better choice:
```yaml
grpc:
  test:
    server:
      port-type: IN_PROCESS # Options: RANDOM_PORT, DEFINED_PORT, IN_PROCESS (default)
```

IN_PROCESS:
```java
@SpringBootTest(classes = InProcessNameIT.Cfg.class)
class InProcessNameIT {

    @InProcessName
    String name;

    @Test
    void testInProcessName() {
        assertThat(name).isNotBlank();
    }

    @Configuration(proxyBeanMethods = false)
    @EnableAutoConfiguration
    static class Cfg {
    }
}
```

DEFINED_PORT:
```java
@SpringBootTest(
        classes = DefinedPortIT.Cfg.class,
        properties = {"grpc.server.port=50054", "grpc.test.server.port-type=DEFINED_PORT"})
class DefinedPortIT {

    @LocalGrpcPort
    int port;

    @Test
    void testDefinedPort() {
        assertThat(port).isEqualTo(50054);
    }

    @Configuration(proxyBeanMethods = false)
    @EnableAutoConfiguration
    static class Cfg {}
}
```

BTW, I think we already have plenty of `@AutoConfigureXxx` annotations. Even as an experienced Spring Boot user, I sometimes find these annotations hard to understand, and it’s even harder for new users. To me, having sensible defaults and configurable options is more appealing than relying on the "magic" of annotations.

**Comment by dsyer** [comment]
I don't think there's any reason we couldn't make in-process the default for tests (with `@SpringBootTest`) and also have an annotation. But the priority was to get the defaults right, and that includes *not* enabling in-process for test main launches. A test annotation isn't a bad way to do that, and we can look at your environment post processor too (but I don't think it will work the way we want with the test main).

**Comment by DanielLiu1123** [comment]
I’m not sure if my understanding of “test main” aligns with yours. I interpret “test main” to mean tests created using `@SpringBootTest`. 

Are you referring to “test main” as directly creating a Spring `ApplicationContext` within the test method instead of using `@SpringBootTest`? For example:
```java
@Test
void testApplicationContext() {
    var ctx = new SpringApplicationBuilder(App.class).run();
    // ...
}
```

If users create the Spring `ApplicationContext` directly within the test method, in-process transport shouldn’t be used by default. This case is similar to web applications, where creating a web `ApplicationContext` within a test method uses the user-defined port instead of a mock environment.

If by “test main” you mean tests created using `@SpringBootTest`, then using an `EnvironmentPostProcessor` should work correctly.

**Comment by dsyer** [comment]
A [test main](https://docs.spring.io/spring-boot/reference/testing/spring-boot-applications.html#testing.spring-boot-applications.using-main) is a main method in a test class (so usually one with `@SpringBootTest`). There are several examples in the samples in this project. Maybe your `EnvironmentPostProcessor` works already with them, I don't know. Someone should check - it would make it quite similar to one of @panic08 's previous drafts.

**Comment by dsyer** [comment]
I kept the annotation and most of the ideas in the `EnvironmentPostProcessor` (the one provided by @panic08 was nice and simple).
